packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1431,25 +1442,28 @@ class TerminalHandler {
   void stop() {
     assert(residentRunner.stayResident);
     for (final MapEntry<io.ProcessSignal, Object> entry in _signalTokens.entries) {
-      globals.signals.removeHandler(entry.key, entry.value);
+      _signals.removeHandler(entry.key, entry.value);
     }
     _signalTokens.clear();
     subscription.cancel();
   }
 
   /// Returns [true] if the input has been handled by this function.
   Future<bool> _commonTerminalInputHandler(String character) async {
-    globals.printStatus(''); // the key the user tapped might be on this line
-    switch(character) {
+    _logger.printStatus(''); // the key the user tapped might be on this line
+    switch (character) {
       case 'a':
         if (residentRunner.supportsServiceProtocol) {
           await residentRunner.debugToggleProfileWidgetBuilds();
           return true;
         }
         return false;
       case 'b':
-        await residentRunner.debugToggleBrightness();
-        return true;
+        if (residentRunner.supportsServiceProtocol) {
+          await residentRunner.debugToggleBrightness();
```

Comment:
Can we push the `supportsServiceProtocol` check into `debugToggleBrightness` and return a bool if it's handled, it it's a requirement?  Otherwise every caller has to know to check it.

Reply:
Some of the methods have meaningful return values, so that would help but it would not necessarily be 100% consistent.

Reply:
More defensive programming would be good in the tool code, we have so many spots where the caller needs to understand the guts and requirements of a method to call it, or things need to be called in a certain order to work...

Reply:
I disagree in general, defensive programming is bad because it makes it harder to track down the cause of crashes. And using crash reporting is just the natural life cycle of a large project that is impossible to completely test.

Though I do agree we should adopt a safer pattern here, and making the invoked function responsible seems like a good adjustment

Reply:
Oh, and as I say that - of course the FlutterDevice class that defines these methods cannot check that 🤦

Reply:
ahh nvm I was being silly